File: packages/ai-engine/src/strategies/direct.ts
/**
 * DirectStrategy — Fast path execution (Graph -> Executor -> Response)
 * Default strategy for simple, low-risk tasks
 *
 * @author Subash Karki
 */
import type {
  ActivationScore,
  ReasoningStrategy,
  StrategyInput,
  StrategyOutput,
  TaskContext,
} from '../types/strategy.js';
import { applyPriorFailurePenalty } from './prior-penalty.js';

export class DirectStrategy implements ReasoningStrategy {
  readonly id = 'direct';
  readonly name = 'Direct Execution';
  readonly version = '1.0.0';
  readonly description =
    'Fast-path strategy for simple, low-risk tasks. Passes graph context straight to the executor.';

  shouldActivate(context: TaskContext): ActivationScore {
    const { complexity, risk } = context;

    let base: ActivationScore;

    if (complexity === 'simple' && risk === 'low') {
      base = { score: 0.9, reason: 'Simple task with low risk — ideal for direct execution' };
    } else if (complexity === 'moderate' && risk === 'low') {
      base = { score: 0.6, reason: 'Moderate complexity but low risk — direct execution viable' };
    } else if (complexity === 'simple' && risk === 'medium') {
      base = { score: 0.5, reason: 'Simple task with medium risk — direct execution possible with caution' };
    } else {
      base = { score: 0.2, reason: 'Complex or high-risk task — direct execution available as fallback' };
    }

    return applyPriorFailurePenalty(base, this.id, context);
  }

  async execute(input: StrategyInput): Promise<StrategyOutput> {
    const start = Date.now();
    const filesUsed = input.graphContext.files.length;

    // Confidence scales with the amount of relevant context available
    const confidence = Math.min(0.95, 0.5 + filesUsed * 0.05);

    // Pass-through: the graph context IS the result for direct strategy
    const result = JSON.stringify({
      type: 'direct-execution',
      goal: input.context.goal,
      files: input.graphContext.files,
      edges: input.graphContext.edges,
    });

    return {
      strategyId: this.id,
      result,
      confidence,
      tokensUsed: 0, // No LLM call — placeholder
      durationMs: Date.now() - start,
      artifacts: {
        strategy: 'direct',
        filesUsed,
      },
    };
  }
}

File: v2/internal/ai/strategies/direct.go
// Package strategies provides AI prompt enhancement strategies.
// DirectStrategy handles simple tasks with minimal enrichment — the message
// passes through unchanged, relying on the existing codebase context.
//
// Author: Subash Karki
package strategies

// DirectStrategy is the default strategy for simple, low-risk tasks.
type DirectStrategy struct{}

// NewDirectStrategy creates a DirectStrategy.
func NewDirectStrategy() *DirectStrategy { return &DirectStrategy{} }

// ID returns the unique identifier for this strategy.
func (d *DirectStrategy) ID() string { return "direct" }

// Name returns a human-readable name.
func (d *DirectStrategy) Name() string { return "Direct Execution" }

// Description returns a brief explanation of what this strategy does.
func (d *DirectStrategy) Description() string {
	return "Fast-path strategy for simple, low-risk tasks. Passes graph context straight to the executor."
}

// ShouldActivate returns a score indicating how well this strategy fits.
func (d *DirectStrategy) ShouldActivate(a TaskAssessment) (float64, string) {
	var score float64
	var reason string
	switch a.Complexity {
	case Simple:
		score, reason = 0.9, "simple task"
	case Moderate:
		score, reason = 0.6, "moderate task"
	default:
		score, reason = 0.3, "complex task, prefer other strategies"
	}
	if a.IsAmbiguous {
		score *= 0.5
		reason += " (penalized: ambiguous requirements)"
	}
	return score, reason
}

// Enrich returns the message unchanged — direct strategy adds no guidance.
func (d *DirectStrategy) Enrich(message string, _ TaskAssessment, _ string) string {
	return message
}
